Modules/_uuidmodule.c

```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+        self->bytes[6] &= 0x0f;  // 0000 1111
+        // Set the variant to RFC 4122/9562 (binary 10xx xxxx)
+        self->bytes[8] |= 0x80;  // 1000 0000
+        // Set the version number (upper 4 bits of byte 6)
+        self->bytes[6] |= (version_num << 4);
+    }
+
+    if (is_safe == Py_None) {
+        // Py_None is immortal; skip decref
+        is_safe = NULL;
+    }
+    if (is_safe != NULL) {
+        Py_INCREF(is_safe);
+        Py_XSETREF(self->is_safe, is_safe);
+    }
+
+    return 0;
+}
+
+
+static const uint8_t INT_TO_HEX[] = "0123456789abcdef";
+
+// Lookup table for hex character to value conversion
+// -1 for invalid characters
+static const int8_t HEX_TO_INT[256] = {
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1, 0,1,2,3,4,5,6,7,8,9,-1,-1,-1,-1,-1,-1,-1,10,11,12,13,14,15,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,10,11,12,13,14,15,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1
+};
+
+static inline void
+byte_to_hex(uint8_t byte, char *hex)
+{
+    hex[0] = INT_TO_HEX[(byte >> 4) & 0xf];
+    hex[1] = INT_TO_HEX[byte & 0xf];
+}
+
+static int
+from_hex(uuidobject *self, PyObject *hex)
+{
+    Py_ssize_t size;
+    const char *start = PyUnicode_AsUTF8AndSize(hex, &size);
+    if (start == NULL) {
+        return -1;
+    }
+
+    uint8_t ch;
+    uint8_t acc, acc_set;
+    int8_t part;
+    int i, j;
+
+    // Reimplement `hex = hex.replace('urn:', '').replace('uuid:', '')`
+    if (size > 0 && start[0] == 'u') {
+        if (size >= 9 && strncmp(start, "urn:uuid:", 9) == 0) {
+            start += 9;
+            size -= 9;
+        }
+        else if (size >= 4 && strncmp(start, "urn:", 4) == 0) {
+            start += 4;
+            size -= 4;
+        }
+        else if (size >= 5 && strncmp(start, "uuid:", 5) == 0) {
+            start += 5;
+            size -= 5;
+        }
+    }
+
+    // Reimplement `hex = hex.strip('{}')`
+    if (size >= 1 && start[0] == '{') {
+        start++;
+        size -= 1;
+    }
+    if (size >= 1 && start[size - 1] == '}') {
+        size -= 1;
+    }
+
+    if (size < 32) {
+        PyErr_SetString(
+            PyExc_ValueError,
+            "badly formed hexadecimal UUID string"
+        );
+        return -1;
+    }
+
+    acc_set = 0;
+    j = 0;
+
+    for (i = 0; i < size; i++) {
+        ch = (uint8_t)start[i];
+
+        if (ch == '-') {
```

Comment:
The exact grammar doesn't allow consecutive `-` in the UUID parsing. It's a known issue that could become hard to fix with a custom logic like that: https://github.com/python/cpython/issues/80938.